src/LinuxTracing/TracerImpl.cpp

```diff
@@ -1125,9 +1125,12 @@ uint64_t TracerImpl::ProcessSampleEventAndReturnTimestamp(const perf_event_heade
       return timestamp_ns;
     }
 
-    UprobesWithStackPerfEvent event = ConsumeUprobeWithStackPerfEvent(ring_buffer, header);
-    DeferEvent(std::move(event));
-    ++stats_.uprobes_with_stack_count;
+    std::optional<UprobesWithStackPerfEvent> event =
+        ConsumeUprobeWithStackPerfEvent(ring_buffer, header);
+    if (event.has_value()) {
```

Comment:
How are the other consume functions handle the case that abi is 0? Do they also return an optional? We might want to be persistent with them.

Reply:
The only other functions to use the regs are the stack sample and the callchain sample functions. I think both of them don't have a case of getting an abi of zero because they only deal with non kernel processes. Even if they get an abi of zero they are moving a regs array so if the array is empty it doesn't cause problems for them. There's also no need to skip those events if abi is zero because there might still be something useful from the event. 

The function that I will add will have a case of abi equals zero and in those functions I will skip the events using optionals. So, in the near future there will be other functions that handle the abi=0 case the same as this one.

Reply:
The fact that abi is never 0, is due to the filtering that already happens at the TracerImpl. It can and does happen also for callstack or callchain samples, and actually we won't be able to produce any useful information in that case. This is the same as for the uprobes+stack samples.

I gave this another thought and think I would argue against this special handling and do instead exactly the same that we also do for callstack/callchain samples:
I.e. keep the perf_event_sample_regs_user_sp struct that contains the abi and allow us to "just move" the data.

This is not ideal, but at least consistent and thus also easier to refactor together at some other point in time.

Reply:
Okay, I removed all of the special handling. There's no optional now. Also, gave the regs_sp struct the same representation as the regs_all struct. Please take a look at all the new changes and tell me what you think.